fe/fe-core/src/test/java/com/starrocks/analysis/RefreshMaterializedViewTest.java

```diff
@@ -746,4 +744,122 @@ public void testMaterializedViewPartitionTTL() throws Exception {
+                "SELECT * FROM mv_union_t2 t2\n");
+
+        String mvName = "mv_union_1";
+        MaterializedView mv = starRocksAssert.getMv("test", "mv_union_1");
+        Assert.assertEquals(
+                Sets.newHashSet("p00010101_20240321", "p20240321_20240322", "p20240322_20240323"),
+                mv.getPartitionNames());
+
+        // cleanup
+        starRocksAssert.dropTable("mv_union_t1");
+        starRocksAssert.dropTable("mv_union_t2");
+        starRocksAssert.dropMaterializedView(mvName);
+    }
+
+    /**
+     * Intersected UNION partition must be same, otherwise will report error
+     */
+    @Test
+    public void testMvOnUnion_IntersectedPartition() throws Exception {
+        starRocksAssert.withTable("CREATE TABLE IF NOT EXISTS mv_union_t1 (\n" +
+                "    leg_id VARCHAR(100) NOT NULL,\n" +
+                "    cabin_class VARCHAR(1) NOT NULL,\n" +
+                "    observation_date DATE NOT NULL\n" +
+                ")\n" +
+                "DUPLICATE KEY(leg_id, cabin_class)\n" +
+                "PARTITION BY RANGE(observation_date) (" +
+                "   PARTITION p1_20240321 VALUES LESS THAN ('2024-03-21'), \n" +
+                "   PARTITION p1_20240322 VALUES LESS THAN ('2024-03-22') \n" +
+                ") ");
+        starRocksAssert.withTable("CREATE TABLE IF NOT EXISTS mv_union_t2 (\n" +
+                "    leg_id VARCHAR(100) NOT NULL,\n" +
+                "    cabin_class VARCHAR(1) NOT NULL,\n" +
+                "    observation_date DATE NOT NULL\n" +
+                ")\n" +
+                "DUPLICATE KEY(leg_id, cabin_class)\n" +
+                "PARTITION BY RANGE(observation_date) (" +
+                "   PARTITION p2_20240322 VALUES LESS THAN ('2024-03-22'), \n" +
+                "   PARTITION p2_20240323 VALUES LESS THAN ('2024-03-23') \n" +
+                ") ");
+        starRocksAssert.withTable("CREATE TABLE IF NOT EXISTS mv_union_t3 (\n" +
+                "    leg_id VARCHAR(100) NOT NULL,\n" +
+                "    cabin_class VARCHAR(1) NOT NULL,\n" +
+                "    observation_date DATE NOT NULL\n" +
+                ")\n" +
+                "DUPLICATE KEY(leg_id, cabin_class)\n" +
+                "PARTITION BY RANGE(observation_date) (" +
+                "   PARTITION p2_20240322 VALUES LESS THAN ('2024-03-21'), \n" +
+                "   PARTITION p2_20240323 VALUES LESS THAN ('2024-04-21') \n" +
+                ") ");
+
+        starRocksAssert.withMaterializedView("CREATE MATERIALIZED VIEW mv1 \n" +
+                "PARTITION BY date_trunc('day', observation_date)\n" +
+                "DISTRIBUTED BY HASH(leg_id)\n" +
+                "REFRESH ASYNC\n" +
+                "AS \n" +
+                "SELECT * FROM mv_union_t1 t1\n" +
+                "UNION ALL\n" +
+                "SELECT * FROM mv_union_t2 t2\n");
+
+        {
+            Exception e = Assert.assertThrows(IllegalArgumentException.class, () ->
```

Comment:
1. Is this a behavior change with previous version? what if refresh mv's all partitions rather than throwing exception?
2. Is that possible to throw exceptions in create mv rather than freshing mv?

Reply:
1. yes and not. previous behavior is not correct, so we have to prohibit it
2. cannot support this scenario. for example, the union mv consists of two children, the first is `[20240102, 20240105)`, the second is `[20240103, 202408)`, you cannot get a correct partition schema for union mv
3. the good news is, it's not a common case, union mv should be used when the partition schemas of base tables are exactly same
4. It's better to check it when creating, I'll try to add a check